### ISSUE TITLE
Test fails as a result of a visible element being deemed not visible.

### ISSUE THREAD

**Issue by todd-m-kemp** [cypress-io/cypress#9378]
### Current behavior
Cypress fails a test indicating that an element is not visible, but the element is actually visible onscreen.

### Desired behavior
The test passes because the element is visible.

### Test code to reproduce
```js
context('Visibility Issue', function () {
  it('Test', function () {
    cy.visit('https://play.vidyard.com/EmrX1jweW4rRGPRJEQeqtn?first_frame=1');
    cy.wait(5000); // Just to make sure everything is loaded.
    cy.contains('Cypress is amazing!')
      .should('be.visible');
  });
});
```
![image](https://user-images.githubusercontent.com/62152998/100639520-44601c80-3303-11eb-8475-acd447c48bd9.png)

The element is clearly visible onscreen, yet this test fails because it says that the element is covered by another element. 🤷‍♂️ 

### Versions

This occurs in Cypress 5.1.0 and 6.0.0 in Chrome 87, Firefox 82, Edge, 87, and Electron 87 on macOS 10.15.7.

**Comment by bahmutov** [comment]
![Screen Shot 2020-11-30 at 12 19 23 PM](https://user-images.githubusercontent.com/2212006/100642127-5db69800-3306-11eb-9d7c-95ee376adb12.png)

That element is really covered by the video player, and it is transparent at this point, but it is impossible for Cypress to really tell this

**Comment by todd-m-kemp** [comment]
>  it is impossible for Cypress to really tell this

Hi @bahmutov ! Thanks for the quick response. Can you elaborate on this a bit more please? I was aware that the element in question was technically covered by the `video` element but since it is transparent I assumed that Cypress would be able to tell that the element "covering" my element was not actually preventing it from being visible, but I guess this was an incorrect assumption. Would you happen to have a suggestion on how to best test this? I would like to make sure that the `Cypress is amazing!` element is visible onscreen, but if that isn't possible using `.should('be.visible')` what would be the next best thing?

**Comment by jennifer-shehane** [comment]
An element being on top of the visible element and transparent does not prevent Cypress from determining its visibility in general. There is something in this case though that we're not taking into account in our visibility algorithm. 

I am able to reproduce the issue with the provided repo, but having a hard time reducing it to the few elements involved. If it's possible to reduce the example any further, that'd be helpful.

**Comment by todd-m-kemp** [comment]
I'm happy to help, but I'm not really sure how to reduce the issue further or provide more details. Is there something in particular that you feel is missing from this that I can try to provide?

**Comment by bahmutov** [comment]
I hoped that https://github.com/cypress-io/cypress/pull/9412 would also fix this issue, but no luck.

**Comment by bahmutov** [comment]
I have described how I debug visibility problems and step through Cypress code in `Cypress.dom.isVisible` using your code as an example. Read https://glebbahmutov.com/blog/debug-visibility/ and/or watch https://www.youtube.com/watch?v=nSmHxxwceDU

**Comment by bartbriek** [comment]
```
describe('Not passing on element should not be visible', () => {
    it('Should pass on shoud not be visible assertion', () => {
        cy.server();
        cy.route({
            method: 'GET',
            url: 'https://api.github.com/repos/cypress-io/cypress'
        }).as('pageLoad');

        // cy.intercept({
        //     method: 'GET',
        //     url: 'https://api.github.com/repos/cypress-io/cypress'
        // }).as('pageLoad');

        cy.visit('www.cypress.io');
        cy.wait('@pageLoad');
        cy.get('.with-bash-effects').should('be.visible');  // npm install cypress button
        cy.get('element_that_is_not_visible').should('not.be.visible');
    });
});
```
In my tests I have assertions similar to this scenario. After a delete action an order row for example should not be visible anymore because it's cancelled. On cypress 5.6.0 this is passing with the cy.server and cy.route.
On cypress 6.1.0 this test fails on the assertion `.should('not.be.visible')` because he can't find the element.

![image](https://user-images.githubusercontent.com/39011473/101896454-52d2f180-3ba9-11eb-93a7-335e03f16632.png)

**Comment by todd-m-kemp** [comment]
Hi @bartbriek .

I don't think this is related.

Cypress 6.0.0 introduced a change to how elements that do not exist respond to the `.should('not.be.visible')` assertion. See the [changelog](https://docs.cypress.io/guides/references/changelog.html#6-0-0) for full details, but the `element_that_is_not_visible` does not exist in the page (as evidenced by the `0` next to it in the screenshot you provided) an therefore the `.should('not.be.visible')` assertion will fail in Cypress 6.1.0. The assertion would not have failed in version 5.6.0 which would explain why it was passing for you previously. After upgrading to Cypress 6.0.0 or above, the assertion that should be used on this element is `.should('not.exist')`.

Hope this helps!

**Comment by christopherost** [comment]
Hi there!
Same problem here regarding the visibility of a <p> with Cypress 7.1.0. Does not occur with Cypress 6.8.0 (which I am currently using). The <p> is said to be covered by a <div> which is used as an event handler listening to click events. It has z-index 1, maybe that is helpful to fix the problem?
Would be glad if I could upgrade to 7.1.0 because 6.8.0 has another issue affecting my tests – but I found a workaround for that, but not for the issue in 7.1.0.

**Comment by Pa6kewi4** [comment]
Hi, I'm new to use cypress, so please don't be hard on me.

The issue is that I can't select Yes or No value in an element, which is hidden until the other element is chosen.

In my image attached, you can see that:
1- If I choose Law Firms, appear next Yes/No element with Id=CompanyDerive_0_0, 
2- below is a script to use a class for choosing the Yes/No value of an element (true or false).
![2022-11-23_16-39-26](https://user-images.githubusercontent.com/118912364/203586718-45033ef5-5c41-4b2e-8ae6-22235f2d2bfa.jpg)

{force: true} is not helping to select a value, cypress continue to give Assertion error that the element is not visible.

Using cypress 11.1.0 end edge browser for testing

**Comment by rooby** [comment]
I had this same problem with Cypress 9.7.0.

I had a test that was doing
`cy.get('.my-element').should('be.visible');`

and it passed always when I run locally with cypress open, but when I run the test in CI it intermittently fails.
When it fails the screenshot clearly shows the element being visible.

I am using Angular, however I did some pretty thorough investigation with MutationObserver to confirm that the element is only added to the DOM once and is never removed after that.
The element also has no animation on it, so it's not fading in or anything like that, and it is never set to 0 width and height, it is just a basic element that once rendered is always visible.

The problem turned out to be (I believe I have a lasting fix) that another element above the element in question was animating in and during its animation must have been overlapping the target element. Even though it wasn't visibly blocking the element due to opacity, Cypress must have thought that it was obscured.

When running locally using cypress open, it was using electron (due to cypress not being able to detect chrome on my work machine), in which case the animation didn't function normally so the test always passed, but in CI when it was using Chrome it would regularly fail.

**Comment by nagash77** [comment]
Hi @todd-m-kemp is this issue still occurring in the latest Cypress?

**Comment by todd-m-kemp** [comment]
We're currently using Cypress 12.8.1 and I am no longer able to reproduce the issue as it is described in the original report. However, I don't know if it's because we changed something about how our video player renders or if the issue was fixed in Cypress. It's been two years since I reported the issue so it's very possible that we changed something on our end that invalidated the original report.
